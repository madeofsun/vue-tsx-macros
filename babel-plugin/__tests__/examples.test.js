const fs = require("node:fs");
const path = require("node:path");
const transformWithPlugin = require("../helpers/transform-with-plugin");

describe("examples", () => {
  test.each(loadSamples())("%s", (_, code) => {
    expect(transformWithPlugin(code, {})).toMatchSnapshot();
  });
});

function loadSamples() {
  const dir = path.resolve(__dirname, "..", "..", "examples");
  return fs
    .readdirSync(dir)
    .filter((fileName) => fileName.endsWith(".ts") || fileName.endsWith(".tsx"))
    .map((fileName) => {
      return [
        fileName,
        fs.readFileSync(path.resolve(dir, fileName)).toString(),
      ];
    });
}
