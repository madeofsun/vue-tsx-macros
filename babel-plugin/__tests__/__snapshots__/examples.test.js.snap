// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`examples functional.tsx 1`] = `
"// literal
export const Example = Object.assign(props => {
  return <div>
      {props.counter.toFixed()}
      {props.color?.trim()}
    </div>;
}, {
  displayName: "Example",
  props: ["color", "counter"]
});
<Example counter={0} color={"red"} />;

// interface
export interface Example1Props {
  color?: "red" | "blue";
  counter: number;
}
export const Example1 = Object.assign(props => {
  return <div>
      {props.counter.toFixed()}
      {props.color?.trim()}
    </div>;
}, {
  displayName: "Example1",
  props: ["color", "counter"]
});
<Example1 counter={0} color={"red"} />;

// type
export type Example2Props = {
  color?: "red" | "blue";
  counter: number;
};
export const Example2 = Object.assign(props => {
  return <div>
      {props.counter.toFixed()}
      {props.color?.trim()}
    </div>;
}, {
  displayName: "Example2",
  props: ["color", "counter"]
});
<Example2 counter={0} color={"red"} />;

// explicit
export const Example3 = Object.assign(props => {
  return <div>{props.color?.trim()}</div>;
}, {
  displayName: "Example3",
  props: {
    color: String
  }
});
<Example3 />;
export default Object.assign(function DefaultExample() {
  return <div />;
}, {
  displayName: "DefaultExample"
});"
`;

exports[`examples functional-defaults.tsx 1`] = `
"const _defaultProps = {
  color: "red",
  array: () => []
};
// reference
export interface ExampleProps {
  counter: number;
  color?: "red" | "blue";
  array?: string[];
}
export const exampleDefaults = {
  color: "red",
  array: () => []
};
export const Example = Object.assign(props => {
  return <div>
        {props.counter.toFixed()}
        {props.color.trim()}
        {props.array.slice()}
      </div>;
}, {
  displayName: "Example",
  props: {
    "counter": {
      type: null,
      default: exampleDefaults["counter"]
    },
    "color": {
      type: null,
      default: exampleDefaults["color"]
    },
    "array": {
      type: null,
      default: exampleDefaults["array"]
    }
  }
});
<Example counter={0} color={"red"} />;

// literal
export const Example1 = Object.assign(props => {
  return <div>
        {props.counter.toFixed()}
        {props.color.trim()}
        {props.array.slice()}
      </div>;
}, {
  displayName: "Example1",
  props: {
    "color": {
      type: null,
      default: _defaultProps["color"]
    },
    "counter": {
      type: null,
      default: _defaultProps["counter"]
    },
    "array": {
      type: null,
      default: _defaultProps["array"]
    }
  }
});
<Example1 counter={0} color={"red"} />;"
`;

exports[`examples functional-emits.tsx 1`] = `
"export type ExampleProps = {
  color?: "red" | "blue";
  counter: number;
};
export type ExampleEmits = {
  change: (value: number) => void;
  update: (value: string) => void;
};
export const Example = Object.assign((props, {
  emit
}) => {
  return <button data-color={props.color?.trim()} onClick={() => {
    emit("change", 0);
    emit("update", "");
  }}>
        {props.counter.toFixed()}
      </button>;
}, {
  displayName: "Example",
  props: ["color", "counter"],
  emits: ["change", "update"]
});
<Example counter={0} onChange={value => value.toFixed()} />;"
`;

exports[`examples standard.tsx 1`] = `
"import { onMounted } from "vue";
// literal
export const Example = {
  name: "Example",
  props: ["color", "counter"],
  setup: props => {
    onMounted(() => {
      console.log("mounted");
    });
    return () => <div>
      {props.counter.toFixed()}
      {props.color?.trim()}
    </div>;
  }
};
<Example counter={0} color={"red"} />;

// interface
export interface Example1Props {
  color?: "red" | "blue";
  counter: number;
}
export const Example1 = {
  name: "Example1",
  props: ["color", "counter"],
  setup: props => {
    onMounted(() => {
      console.log("mounted");
    });
    return () => <div>
      {props.counter.toFixed()}
      {props.color?.trim()}
    </div>;
  }
};

// type
export type Example2Props = {
  color?: "red" | "blue";
  counter: number;
};
export const Example2 = {
  name: "Example2",
  props: ["color", "counter"],
  setup: props => {
    onMounted(() => {
      console.log("mounted");
    });
    return () => <div>
      {props.counter.toFixed()}
      {props.color?.trim()}
    </div>;
  }
};

// ComponentPropsOptions
export const Example3 = {
  name: "Example3",
  props: {
    color: String
  },
  setup: props => {
    onMounted(() => {
      console.log("mounted");
    });
    return () => <div>{props.color?.trim()}</div>;
  }
};

// name from function
export default {
  name: "DefaultExample",
  setup: function DefaultExample() {
    onMounted(() => {
      console.log("mounted");
    });
    return () => <div />;
  }
};"
`;

exports[`examples standard-defaults.tsx 1`] = `
"const _defaultProps = {
  color: "red",
  array: () => []
};
// reference
export interface ExampleProps {
  counter: number;
  color?: "red" | "blue";
  array?: string[];
}
export const exampleDefaults = {
  color: "red",
  array: () => []
};
export const Example = {
  name: "Example",
  props: {
    "counter": {
      type: null,
      default: exampleDefaults["counter"]
    },
    "color": {
      type: null,
      default: exampleDefaults["color"]
    },
    "array": {
      type: null,
      default: exampleDefaults["array"]
    }
  },
  setup: props => {
    return () => <div>
        {props.counter.toFixed()}
        {props.color.trim()}
        {props.array.slice()}
      </div>;
  }
};
<Example counter={0} color={"red"} />;

// literal
export const Example1 = {
  name: "Example1",
  props: {
    "color": {
      type: null,
      default: _defaultProps["color"]
    },
    "counter": {
      type: null,
      default: _defaultProps["counter"]
    },
    "array": {
      type: null,
      default: _defaultProps["array"]
    }
  },
  setup: props => {
    return () => <div>
        {props.counter.toFixed()}
        {props.color.trim()}
        {props.array.slice()}
      </div>;
  }
};
<Example1 counter={0} color={"red"} />;"
`;

exports[`examples standard-emits.tsx 1`] = `
"export type ExampleProps = {
  color?: "red" | "blue";
  counter: number;
};
export type ExampleEmits = {
  change: (value: number) => void;
  update: (value: string) => void;
};
export const Example = {
  name: "Example",
  props: ["color", "counter"],
  emits: ["change", "update"],
  setup: (props, {
    emit
  }) => {
    return () => <button data-color={props.color?.trim()} onClick={() => {
      emit("change", 0);
      emit("update", "");
    }}>
        {props.counter.toFixed()}
      </button>;
  }
};
<Example counter={0} onChange={value => value.toFixed()} />;"
`;

exports[`examples standard-expose.tsx 1`] = `
"let _render;
import { ref } from "vue";
export const Example = {
  name: "Example",
  setup: () => {
    const counter = ref(0);
    const increment = () => counter.value += 1;
    _render = () => <button onClick={increment}>{counter.value}</button>;
    return {
      increment
    };
  },
  render() {
    return _render();
  }
};
const instance = ((null as unknown) as InstanceType<typeof Example>);
console.log(instance.increment().toFixed());"
`;
